### DISCUSSION TITLE
What's the maximum tsconfig target I can use for a library used in Next.js?

### DISCUSSION THREAD

**Discussion by will-stone** [vercel/next.js#40949] [Help]
If I am building a component or utility library to be used in a Next.js application, that **doesn't** need to support IE11, what's the maximum target I can use? `es5` / `es2015` / `...` / `esnext`? Will Next reject importing anything higher than `es5`? Or does it not care and we can use anything as long as it supports the browsers we'd like to support?

Thank you

**Comment by icyJoseph** [comment]
## TL;DR

`ES2022` as of the time of writing, though `ES6` (`ES2015`) is perhaps, when excluding IE11, the safest bet.

Not sure if I overcomplicated my analysis of this, or if its just a complex subject with many caveats, but here it goes anyway. I'll keep on reading, testing and editing the answer as necessary, but I think I am kind of happy with it, for now anyway. Hope to see others take on it, and rebuttals to my line of thought here.

## Here we go...

With IE11 out of the support scope of the project, I think ES2015, because of `arrow functions`, but at the cost `async/await` readability. 

It might be more important to use `CommonJS/ESM` properly in your utility library.

Though you could certainly use higher versions, I expand on that down below, `esnext` is interesting because it changes with the TypeScript version you have installed, which in some cases might not be desirable. For example if a feature becomes available in the type system, but the node version running your code doesn't support it yet.

### Arrow functions

My reasoning is that since IE is out of the way, then you don't need to transform from arrow functions `() => {}` to regular functions `function(){}`.

### async/await

A problem with older targets is  that  `async/await` is transformed to use generators and Promises, and not kept as is. 

I think ES2017 is needed to keep `async/await` as is. For `ES3` and `ES5`, `async/await` requires even more complex runtime supporting code.

In either case `async/await` would bare some overhead, and it'd make your compiled code harder to read when debugging, but it ensure broader browser coverage.

### Higher targets

Using things higher than `ES5`, or `ES2015`, or `ES0217`, is all about what kind of browser support you want to give. 

Say if we allow `.at` which exists since Chrome 92, but our server runs Node 14, where it doesn't exist, we might have problems with code that runs on both platforms.

This is more of developer ergonomics than a runtime guarantee. If I try to access `.at` but my `target` is not high enough, the type system yells at me.

Perhaps the same can be said about decorators, though I haven't worked so much with those.

###  Will Next reject importing anything higher than es5?

Take a look at [RFC: Change default JS/CSS output to target modern browsers](https://github.com/vercel/next.js/discussions/33227)

> The above is interesting because code written within the Next.js app, as far as I understand, will be compiled to ES5, at the time of writing.

It is up to the browser to support the methods we've used. Also at runtime, on the server, if the runtime ever hits a branch of code that's using methods not present in the node version, it'll crash. 

Next.js has no real saying, other than through TypeScript definitions, and proper module resolutions (CommonJS vs ESM), on the code you bring in. If it can be parsed, compiled and ran by the Browser and Node.js, it'd be so.

So if my utils library used `ES2022` with `async/await`, when I bring it into my app, Next.js uses it as it was compiled.

For example if my utility library is:

```ts
export const sleep = async (time: number) => {
  await new Promise((next) => setTimeout(next, time, null));
  return time;
};
```

When compiled using `ES2022` and `ESM`, it stays pretty much the same, minus the types, I then install it, and use it. 

While developing, I cannot pass strings to `sleep`, TS disallows that, and then when everything is correct, and I build the application, I can find the utility code inlined with the rest of the application. 

```js
      const a = async (n) => (
        await new Promise((t) => setTimeout(t, n, null)), n
      );
    // or if I used CommonJS
    t._ = async (n) => (await new Promise((t) => setTimeout(t, n, null)), n);
```

I guess the exact output changes and depends on compiler/bundler optimisations, but at the core, the `async/await` syntax prevails. So this code better not run on browser versions that do not support `async/await` as is, which is why I think using `ES2015` in the utility library, is a good compromise, when IE11 is to be ignored.

### TypeScript versions

I guess that differences in the typescript versions used by the utility package and the consumer, could cause issues, like if the utility package uses `Awaited` not it is present in the consumer side. Not 100% sure though. 

## Sources

As always take what you read from others online with a grain of salt, there's nothing better than testing yourself and reading the official docs:

- https://www.typescriptlang.org/tsconfig#lib
- https://www.typescriptlang.org/tsconfig#target
- https://www.typescriptlang.org/play
- https://github.com/vercel/next.js/discussions/33227

**Reply by will-stone** [reply]
Thank you @icyJoseph, `ES2015` is enough for now, as I'm basically trying to avoid `downlevelIteration` 😄

**Reply by itzomen** [reply]
Great answer, thanks
I'm gonna go with es2017
